optuna/_convert_positional_args.py

```diff
@@ -37,32 +48,73 @@ def convert_positional_args(
     *,
     previous_positional_arg_names: Sequence[str],
     warning_stacklevel: int = 2,
+    deprecated_version: str | None = None,
+    removed_version: str | None = None,
 ) -> "Callable[[Callable[_P, _T]], Callable[_P, _T]]":
     """Convert positional arguments to keyword arguments.
 
     Args:
-        previous_positional_arg_names: List of names previously given as positional arguments.
-        warning_stacklevel: Level of the stack trace where decorated function locates.
+        previous_positional_arg_names:
+            List of names previously given as positional arguments.
+        warning_stacklevel:
+            Level of the stack trace where decorated function locates.
+        deprecated_version:
+            The version in which the use of positional arguments is deprecated.
+        removed_version:
+            The version in which the use of positional arguments will be removed.
     """
 
+    if deprecated_version is None and removed_version is None:
+        pass
+    else:
+        if deprecated_version is None:
+            raise ValueError(
+                "deprecated_version must not be None when removed_version is specified."
+            )
+        if removed_version is None:
+            raise ValueError(
+                "removed_version must not be None when deprecated_version is specified."
+            )
+
+        _validate_version(deprecated_version)
+        _validate_version(removed_version)
+        _validate_two_version(deprecated_version, removed_version)
+
     def converter_decorator(func: "Callable[_P, _T]") -> "Callable[_P, _T]":
+
         assert set(previous_positional_arg_names).issubset(set(signature(func).parameters)), (
             f"{set(previous_positional_arg_names)} is not a subset of"
             f" {set(signature(func).parameters)}"
         )
 
         @wraps(func)
         def converter_wrapper(*args: Any, **kwargs: Any) -> "_T":
+            warning_messages = []
             positional_arg_names = _get_positional_arg_names(func)
             inferred_kwargs = _infer_kwargs(previous_positional_arg_names, *args)
+
+            if inferred_kwargs and (deprecated_version or removed_version):
```

Comment:
`inferred_kwargs` include ordinal positional arguments. So we should check the lengths of them. 

```suggestion
            if len(inferred_kwargs) > len(positional_arg_names) and (deprecated_version or removed_version):
```